dexmaker/src/main/java/com/android/dx/AnnotationId.java

```diff
@@ -0,0 +1,166 @@
+        this.declaringType = declaringType;
+        this.type = type;
+        this.annotationType = annotationType;
+        this.elements = new ArrayList<>();
+    }
+
+    /**
+     *  Generate an annotation.
+     *
+     * @param declaringType the type declaring the method that this annotation is added to.
+     * @param type the type of value this annotation holds.
+     * @param annotationType The type of this annotation which is for method, field or class.
+     * @return an annotation {@code AnnotationId<D,V>}
+     */
+    public static <D, V> AnnotationId<D, V> get(TypeId<D> declaringType, TypeId<V> type,
+                                                ElementType annotationType) {
+        return new AnnotationId<>(declaringType, type, annotationType);
+    }
+
+    /**
+     * Add an element to this annotation.
+     *
+     * @param element annotation element to be added.
+     */
+    public void addElement(Element element) {
+        CstString pairName = new CstString(element.getName());
+        Constant pairValue = Element.toConstant(element.getValue());
+        NameValuePair nameValuePair = new NameValuePair(pairName, pairValue);
+        elements.add(nameValuePair);
+    }
+
+    /**
+     * Add this annotation to a method.
+     *
+     * @param dexMaker DexMaker instance.
+     * @param method Method to be added to.
+     */
+    public void addToMethod(DexMaker dexMaker, MethodId<?, ?> method) {
+        if (annotationType != ElementType.METHOD) {
+            throw new IllegalStateException("This annotation is not for method");
+        }
+
+        if (method.declaringType != declaringType) {
+            throw new IllegalArgumentException("Method" + method + "'s declaring type is inconsistent with" + this);
+        }
+
+        ClassDefItem classDefItem = dexMaker.getTypeDeclaration(declaringType).toClassDefItem();
+
+        if (classDefItem == null) {
+            throw new NullPointerException("No class defined item is found");
+        } else {
+            CstMethodRef cstMethodRef = method.constant;
+
+            if (cstMethodRef == null) {
+                throw new NullPointerException("Method reference is NULL");
+            } else {
+                // Generate CstType
+                CstType cstType = CstType.intern(type.ropType);
+
+                // Generate Annotation
+                Annotation annotation = new Annotation(cstType, AnnotationVisibility.RUNTIME);
+
+                // Add generated annotation
+                Annotations annotations = new Annotations();
+                for (NameValuePair nvp : elements) {
+                    annotation.add(nvp);
+                }
+                annotations.add(annotation);
+                classDefItem.addMethodAnnotations(cstMethodRef, annotations, dexMaker.getDexFile());
+            }
+        }
+    }
+
+    /**
+     *  Annotation Element
+     */
+    public static final class Element {
```

Comment:
Can we get some more complete documentation on this class? It looks like it's something you expect consumers to use directly, so it should have docs that make it clear what it's for and how to use it.

Reply:
Yes, I will add more documentation.

Reply:
It's done by new commit 733f075